File: typescript/core/test/retry.test.ts
/**
 * Retry-with-exponential-backoff around the model call. Driven by the reusable
 * `MockLlmProvider` (it can script transient errors via `pushError`). Backoff is set
 * to 0 so no real time is spent sleeping.
 */

import { describe, expect, it } from 'vitest';
import { AgentOptions, SmoothAgent } from '../src/agent.js';
import { MockLlmProvider } from '../src/llmProvider.js';

function makeAgent(client: MockLlmProvider, options: AgentOptions = {}): SmoothAgent {
    return new SmoothAgent(client, options);
}

describe('SmoothAgent retry', () => {
    it('retries then succeeds', async () => {
        // Errors k times then a text reply; maxRetries >= k → the turn succeeds and the
        // model is called exactly k+1 times.
        const client = new MockLlmProvider();
        client.pushError('rate limited').pushError('rate limited').pushText('ok');
        const agent = makeAgent(client, { maxRetries: 2, retryBackoffMs: 0 });
        const result = await agent.run('hi');
        expect(result.text).toBe('ok');
        expect(client.callCount).toBe(3); // k+1 = 2 failures + 1 success
    });

    it('propagates the error when retries are exhausted', async () => {
        // Errors maxRetries+1 times → the provider error propagates (the turn fails).
        const client = new MockLlmProvider();
        client.pushError('boom').pushError('boom');
        const agent = makeAgent(client, { maxRetries: 1, retryBackoffMs: 0 });
        await expect(agent.run('hi')).rejects.toThrow('boom');
        expect(client.callCount).toBe(2); // maxRetries + 1 attempts
    });

    it('does not retry by default (maxRetries=0)', async () => {
        // Default maxRetries=0 → a single error propagates immediately (one attempt).
        const client = new MockLlmProvider();
        client.pushError('nope').pushText('never reached');
        const agent = makeAgent(client, {});
        await expect(agent.run('hi')).rejects.toThrow('nope');
        expect(client.callCount).toBe(1);
    });
});

File: go/core/retry_test.go
package core

import (
	"context"
	"strings"
	"testing"
)

// Retry-with-exponential-backoff around the model call, driven by the reusable
// MockLlmProvider (it can script a transient error via PushError). RetryBackoff is
// left at its zero value so no real time is spent sleeping.

func TestRetriesThenSucceeds(t *testing.T) {
	// Errors k times then a text reply; MaxRetries >= k → the turn succeeds and the
	// model is called exactly k+1 times.
	mock := NewMockLlmProvider()
	mock.PushError("rate limited").PushError("rate limited").PushText("ok")
	agent := NewSmoothAgent(mock, AgentOptions{MaxRetries: 2})
	res, err := agent.Run(context.Background(), "hi", nil)
	if err != nil {
		t.Fatalf("expected success, got error: %v", err)
	}
	if res.Text != "ok" {
		t.Fatalf("want text %q, got %q", "ok", res.Text)
	}
	if mock.CallCount() != 3 { // k+1 = 2 failures + 1 success
		t.Fatalf("want 3 model calls, got %d", mock.CallCount())
	}
}

func TestErrorPropagatesWhenRetriesExhausted(t *testing.T) {
	// Errors MaxRetries+1 times → the provider error propagates (the turn fails).
	mock := NewMockLlmProvider()
	mock.PushError("boom").PushError("boom")
	agent := NewSmoothAgent(mock, AgentOptions{MaxRetries: 1})
	_, err := agent.Run(context.Background(), "hi", nil)
	if err == nil || !strings.Contains(err.Error(), "boom") {
		t.Fatalf("want error containing %q, got %v", "boom", err)
	}
	if mock.CallCount() != 2 { // MaxRetries + 1 attempts
		t.Fatalf("want 2 model calls, got %d", mock.CallCount())
	}
}

func TestNoRetryByDefault(t *testing.T) {
	// Default MaxRetries=0 → a single error propagates immediately (one attempt).
	mock := NewMockLlmProvider()
	mock.PushError("nope").PushText("never reached")
	agent := NewSmoothAgent(mock, AgentOptions{})
	_, err := agent.Run(context.Background(), "hi", nil)
	if err == nil || !strings.Contains(err.Error(), "nope") {
		t.Fatalf("want error containing %q, got %v", "nope", err)
	}
	if mock.CallCount() != 1 {
		t.Fatalf("want 1 model call, got %d", mock.CallCount())
	}
}
